### PR TITLE
Account Recovery Token Getter and Create New Codes

### PR DESCRIPTION
**Pull request by kimlisa**
part of RFD 0029: https://github.com/gravitational/teleport/blob/master/rfd/0029-account-lifecycle.md
split from larger PR for easier review: https://github.com/gravitational/teleport/pull/7405

### Description
- Implement `GetAccountRecoveryToken`: validates and retrieves a recovery token
- Implement `CreateRecoveryCodes`: creates new recovery codes for users who meet requirements, invalidating their previous existing codes

### CODE REVIEW

File: api/client/proto/authservice.proto [hunk 1]
```diff
@@ -1094,6 +1094,31 @@ message CompleteAccountRecoveryRequest {
     }
 }
 
+// CreateAccountRecoveryCodesRequest is a request to create new set of recovery codes for a user,
+// replacing and invalidating any previously existing codes. Recovery codes can only be given to
+// users who meet the following requirements:
+//  - cloud feature is enabled
+//  - username is in valid email format
+message CreateAccountRecoveryCodesRequest {
+    // TokenID is the ID of a user token that will be used to verify this request.
```
Comment (codingllama) [inline api/client/proto/authservice.proto:1103]: Does it have to be a specific kind of token?

File: lib/auth/accountrecovery.go [hunk 2]
```diff
@@ -451,6 +451,60 @@ func (s *Server) CompleteAccountRecovery(ctx context.Context, req *proto.Complet
 	return nil
 }
 
+// CreateAccountRecoveryCodes implements AuthService.CreateAccountRecoveryCodes.
+func (s *Server) CreateAccountRecoveryCodes(ctx context.Context, req *proto.CreateAccountRecoveryCodesRequest) (*proto.CreateAccountRecoveryCodesResponse, error) {
+	errMsg := "unable to create new recovery codes, please contact your system administrator"
```
Comment (codingllama) [inline lib/auth/accountrecovery.go:456]:
```suggestion
	const unableToCreateCodesMsg = "unable to create new recovery codes, please contact your system administrator"
```

File: api/client/proto/authservice.proto [hunk 3]
```diff
@@ -1468,4 +1493,16 @@ service AuthService {
...
+    // user, replacing and invalidating any previously owned codes. Users can only get recovery
+    // codes if their username is in a valid email format.
+    rpc CreateAccountRecoveryCodes(CreateAccountRecoveryCodesRequest)
+        returns (CreateAccountRecoveryCodesResponse);
+    // GetAccountRecoveryToken (exclusive to cloud users) returns a user token resource after
+    // verifying that the token requested has not expired and is of the correct recovery kind.
+    // Besides checking for validity of a token ID, it is also used to get basic information from
+    // the token e.g: username, state of recovery (started or approved) and the type of recovery
+    // requested (password or second factor).
+    rpc GetAccountRecoveryToken(GetAccountRecoveryTokenRequest) returns (types.UserTokenV3);
```
Comment (codingllama) [inline api/client/proto/authservice.proto:1507]: Discussed offline, but registering some of the thoughts here.

It looks like someone could try to use this to guess valid recovery tokens. There are some reasons for the RPC to exist (mostly for better UX), but personally I'd rather not have it at all - if the user loses the token for some reason during recovery (closed browser or whatever), then they should start over. One could argue that the circumstances for guessing are rare, and the time window is tight, but I'd rather not have to consider this at all.

As a compromise for keeping it, let's make sure the RPC is rate limited, so at least it is a bit harder to attack.

File: lib/auth/accountrecovery_test.go [hunk 4]
```diff
@@ -1060,6 +1060,156 @@ func TestAccountRecoveryFlow(t *testing.T) {
 	}
 }
 
+func TestGetAccountRecoveryToken(t *testing.T) {
+	t.Parallel()
+	srv := newTestTLSServer(t)
+
+	cases := []struct {
+		name     string
+		isExpErr bool
```
Comment (codingllama) [inline lib/auth/accountrecovery_test.go:1069]:
```suggestion
		wantErr bool
```

File: lib/auth/accountrecovery_test.go [hunk 5]
```diff
@@ -1060,6 +1060,156 @@ func TestAccountRecoveryFlow(t *testing.T) {
...
+				TTL:  5 * time.Minute,
+				Type: UserTokenTypeResetPassword,
+			},
+		},
+		{
+			name:     "token not found",
+			isExpErr: true,
+			req: CreateUserTokenRequest{
+				TTL:  5 * time.Minute,
+				Type: "Type-Does-Not-Exist",
```
Comment (codingllama) [inline lib/auth/accountrecovery_test.go:1085]: Maybe:

```suggestion
				Type: "unknown-token-type",
```

so it looks less like an impossible situation and more like future proofing?

File: lib/auth/accountrecovery_test.go [hunk 6]
```diff
@@ -1060,6 +1060,156 @@ func TestAccountRecoveryFlow(t *testing.T) {
...
+				Name: "llama@email.com",
+				TTL:  5 * time.Minute,
+				Type: UserTokenTypeResetPassword,
+			},
+		},
+		{
+			name: "token not found",
+			req: CreateUserTokenRequest{
+				TTL:  5 * time.Minute,
+				Type: "Type-Does-Not-Exist",
```
Comment (codingllama) [inline lib/auth/accountrecovery_test.go:1164]:
```suggestion
				Type: "unknown-token-type",
```

File: lib/auth/accountrecovery_test.go [hunk 7]
```diff
@@ -1060,6 +1060,156 @@ func TestAccountRecoveryFlow(t *testing.T) {
...
+		SecondFactor: constants.SecondFactorOTP,
+	})
+	require.NoError(t, err)
+
+	err = srv.Auth().SetAuthPreference(ctx, ap)
+	require.NoError(t, err)
+
+	cases := []struct {
+		name     string
+		isNotErr bool
```
Comment (codingllama) [inline lib/auth/accountrecovery_test.go:1149]: Suggestion: Remove negation from the name? Makes it a bit confusing to read.

```suggestion
		wantErr bool
```

File: lib/auth/accountrecovery_test.go [hunk 8]
```diff
@@ -1060,6 +1060,156 @@ func TestAccountRecoveryFlow(t *testing.T) {
...
+		token, err := srv.Auth().newUserToken(c.req)
+		require.NoError(t, err)
+		_, err = srv.Auth().Identity.CreateUserToken(context.Background(), token)
+		require.NoError(t, err)
+
+		switch {
+		case c.isNotErr:
+			res, err := srv.Auth().CreateAccountRecoveryCodes(ctx, &proto.CreateAccountRecoveryCodesRequest{
+				TokenID: token.GetName(),
+			})
```
Comment (codingllama) [inline lib/auth/accountrecovery_test.go:1196]: Pull this call out of the if? This is the thing we are testing, plus it happens in both branches.

File: lib/auth/accountrecovery_test.go [hunk 9]
```diff
@@ -1060,6 +1060,156 @@ func TestAccountRecoveryFlow(t *testing.T) {
...
+			require.NoError(t, err)
+			require.Len(t, res.GetRecoveryCodes(), 3)
```
Comment (codingllama) [inline lib/auth/accountrecovery_test.go:1198]: nit: I would prefer a non-exact check - the number of codes could change (increase, in particular) and this test wouldn't necessarily have to break. Maybe >1 or >=3?.  Feel free to disregard.

**Review by codingllama — CHANGES_REQUESTED** [review]
Requested changes for the rate limiting (see comments), otherwise looks good.

File: api/client/proto/authservice.proto (hunk 1 above)
Comment (kimlisa) [reply api/client/proto/authservice.proto:1103]: I left it open to accept `privilege token` for later.

Comment (codingllama) [reply api/client/proto/authservice.proto:1103]: Should we add that to the docs?

File: api/client/proto/authservice.proto (hunk 3 above)
Comment (kimlisa) [reply api/client/proto/authservice.proto:1507]: added rate limit

File: lib/auth/accountrecovery_test.go (hunk 9 above)
Comment (kimlisa) [reply lib/auth/accountrecovery_test.go:1198]: i used the const value

**Review by codingllama — APPROVED** [review]
LGTM.

**Review by klizhentas — APPROVED** [review]

**Pull request merged**